### ISSUE TITLE
possible problem with order of plugin import

### ISSUE THREAD

**Issue by bruceravel** [xraypy/xraylarch#116]
Hi Matt,

This morning, with a freshly `git pull`-ed larch, I am unable to start the larch interpreter.

Below is the traceback along with the output of these print statements

``` python
                print "\n", mod, filelist
                for fname in filelist:
                    print fname, fname[:-3]
```

at this location: https://github.com/xraypy/xraylarch/blob/master/lib/builtins.py#L393

```
/home/bruce/.larch/plugins/wx ['scanviewer.py', 'periodictable.py', 'gui_utils.py', 'xrfdisplay_fitpeaks.py', 'plotter.py', 'xrfdisplay.py', 'mapviewer.py', '__init__.py', 'gse_dtcorrect.py', 'xrfdisplay_utils.py', 'parameter.py', 'mapimageframe.py']
scanviewer.py scanviewer
periodictable.py periodictable
gui_utils.py gui_utils
xrfdisplay_fitpeaks.py xrfdisplay_fitpeaks
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 411, in _plugin_file
    out = imp.load_module(plugin, fh, modpath, desc)
  File "/home/bruce/.larch/plugins/wx/xrfdisplay_fitpeaks.py", line 24, in <module>
    from parameter import ParameterPanel
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 397, in _plugin_file
    ret =  _plugin_file(fname[:-3], path=[mod])
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 419, in _plugin_file
    write(traceback.print_tb(tback))
Traceback (most recent call last):
  File "/usr/local/bin/larch", line 99, in <module>
    shell = larch.shell(banner_msg=BANNER, quiet=options.quiet)
  File "/usr/local/lib/python2.7/dist-packages/larch/shell.py", line 53, in __init__
    self.larch  = Interpreter()
  File "/usr/local/lib/python2.7/dist-packages/larch/interpreter.py", line 142, in __init__
    builtins._addplugin(pdir, _larch=self)
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 443, in _addplugin
    return _plugin_file(plugin)
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 400, in _plugin_file
    write(traceback.print_tb(tback))
TypeError: expected a character buffer object
```

The location of the beginning of the traceback seems to be the line in `plugins/wx/xrfdisplay_fitpeaks.py` that says

``` python
from parameter import ParameterPanel
```

and the problem seems to be that `parameter.py` has not been imported yet (it comes later in `filelist`).

I can make larch finish startup by manually prepending `parameter.py` to `filelist`.

Do you need to maintain a dependence tree so that plugins are imported in the right order?

**Comment by newville** [comment]
Yuck.  Will investigate.

On Thu, Apr 30, 2015 at 8:48 AM, Bruce Ravel <notifications@github.com>
wrote:

> Hi Matt,
>
> This morning, with a freshly git pull-ed larch, I am unable to start the
> larch interpreter.
>
> Below is the traceback along with the output of these print statements
>
>                 print "\n", mod, filelist
>                 for fname in filelist:
>                     print fname, fname[:-3]
>
> at this location:
> https://github.com/xraypy/xraylarch/blob/master/lib/builtins.py#L393
>
> /home/bruce/.larch/plugins/wx ['scanviewer.py', 'periodictable.py', 'gui_utils.py', 'xrfdisplay_fitpeaks.py', 'plotter.py', 'xrfdisplay.py', 'mapviewer.py', '__init__.py', 'gse_dtcorrect.py', 'xrfdisplay_utils.py', 'parameter.py', 'mapimageframe.py']
> scanviewer.py scanviewer
> periodictable.py periodictable
> gui_utils.py gui_utils
> xrfdisplay_fitpeaks.py xrfdisplay_fitpeaks
>   File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 411, in _plugin_file
>     out = imp.load_module(plugin, fh, modpath, desc)
>   File "/home/bruce/.larch/plugins/wx/xrfdisplay_fitpeaks.py", line 24, in <module>
>     from parameter import ParameterPanel
>   File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 397, in _plugin_file
>     ret =  _plugin_file(fname[:-3], path=[mod])
>   File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 419, in _plugin_file
>     write(traceback.print_tb(tback))
> Traceback (most recent call last):
>   File "/usr/local/bin/larch", line 99, in <module>
>     shell = larch.shell(banner_msg=BANNER, quiet=options.quiet)
>   File "/usr/local/lib/python2.7/dist-packages/larch/shell.py", line 53, in __init__
>     self.larch  = Interpreter()
>   File "/usr/local/lib/python2.7/dist-packages/larch/interpreter.py", line 142, in __init__
>     builtins._addplugin(pdir, _larch=self)
>   File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 443, in _addplugin
>     return _plugin_file(plugin)
>   File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 400, in _plugin_file
>     write(traceback.print_tb(tback))
> TypeError: expected a character buffer object
>
> The location of the beginning of the traceback seems to be the line in
> plugins/wx/xrfdisplay_fitpeaks.py that says
>
> from parameter import ParameterPanel
>
> and the problem seems to be that parameter.py has not been imported yet
> (it comes later in filelist).
>
> I can make larch finish startup by manually prepending parameter.py to
> filelist.
>
> Do you need to maintain a dependence tree so that plugins are imported in
> the right order?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xraypy/xraylarch/issues/116>.
>



-- 
--Matt Newville <newville at cars.uchicago.edu> 630-252-0431

**Comment by bruceravel** [comment]
It may be useful to consider doing some kind of formal graph of dependencies between plugins.  (Say, something like https://bitbucket.org/blais/snakefood .)

It is easy to imagine a situation where someone makes `a.py` depend upon `b.py` even though `b.py` already depends upon `a.py`.  If you had a graph of plugin dependencies, this problem wouldn't happen and you could test for cycles (and issue an understandable error message) before starting plugin import.

**Comment by newville** [comment]
Oddly, I cannot reproduce this problem on Linux or Windows (all I've tried).     

But, there is definitely an error in printing out the tracebacks for Import Errors with plugin code.  Commit  627312c911a8d9ee91390e92a303124ecf43d1be  should fix *that*.     But I think that might get you just to the point of being able to better diagnose the problem.   FWIW, I'm not sure I love the effect of this change -- an error in any plugin will stop the automatic import of all subsequent plugins (and the order is sort of accidental).   

So, it could be that some sort of circular import is at fault, but it's a little surprising to me that this would show up at plugins/wx/parameter.py,  as it is clean for importing other plugins.  

I do see that it does import wxutls (which I think, like wxmplot is not sufficiently checked for as a dependency).    

The snakefood link looks like a good utility that should be used.

**Comment by bruceravel** [comment]
I'm not surprised that you haven't reproduced this.  Had you seen it, you would have fixed is :smile: 

I think that the order or items returned by `os.listdir()` is unpredictable.

I have seen this behavior on two computers, both ubuntu, one 14.04, one 14.10.

As I said, the problem went away when I manually pre-pended `parameter.py` to the list.  I don't think circular import, per se, is the problem.  Just out-of-order import.  I bet, on your computers, for some reason, `os.listdir()` returns `parameter.py` before `xrfdisplay_fitpeaks.py`.

**Comment by newville** [comment]
Yep, you're right.  

I'll add that fix for `parameter.py` and `xrfdisplay_fitpeaks.py`,  but also sort the result of `os.listdir()` so the order more predictable between systems and problems more easily found.

**Comment by bruceravel** [comment]
Finally back at my desk and with enough bandwidth to look at this...

What you did in https://github.com/xraypy/xraylarch/commit/627312c911a8d9ee91390e92a303124ecf43d1be results in a much better error message, which is great, and makes it to the larch prompt, also great.  But the problem persists.  Here's the current error message, fyi:

```
bruce@titanium [xraylarch](master) > larch
  Larch 0.9.24 (28-Oct-2014) M. Newville and the Larch Development Team
  using python 2.7.6, numpy 1.9.1, wx-enabled, wx version 2.8.12.1
  File "/usr/local/lib/python2.7/dist-packages/larch/builtins.py", line 421, in _plugin_file
    out = imp.load_module(plugin, fh, modpath, desc)
  File "/home/bruce/.larch/plugins/wx/xrfdisplay_fitpeaks.py", line 24, in <module>
    from parameter import ParameterPanel
Python Error at plugin '/home/bruce/.larch/plugins/wx/xrfdisplay_fitpeaks.py', line 0
   ^
ImportError: No module named parameter
larch>
```

**Comment by newville** [comment]
should be fixed at 475ca99fe1d3da5bdbd7c9e267174e5a29bbced7

**Comment by bruceravel** [comment]
it is :+1:  

however, if I were to make a PR changing `parameters.py` to `zarameters.py` ....

**Comment by newville** [comment]
That would actually be OK (and is, of course, now very tempting!), in the sense that it would never work  unless the import in `wx/__init__.py` were changed to 

``` python
from .zarameters import ParameterPanel
```
in which case it would always work.    That is, the change in `xrfdisplay_fitpeaks.py` to 

```python
from larch_plugins.wx import ParameterPanel
```
requires that to be explicitly defined in `wx/__init__.py`.  Before it was implicit (and actually vague -- had there been a `parameter.py` module somewhere higher in the path, that could have been found and caused a similar problem).     The recent move to `from larch_plugins.XXX import Thing`  may still be fresh enough that we're putting out the last of the fires.     

There may be an errant import or two left -- `use_plugin_path` is still left in a few modules.  But  sorting the order of listdir() is a good idea anyway....  Plus,  I added a list of "core plugins" to always load first.   So, I think it may be time to fully get rid of `use_plugin_path`.
